File: coding/src/com/algos/li09_linkedlist/revised/L1472_DesignBrowserHistory.java
package com.algos.li09_linkedlist.revised;

public class L1472_DesignBrowserHistory {
    public static void main(String[] args) {
        L1472_DesignBrowserHistory obj = new L1472_DesignBrowserHistory();
        BrowserHistory browserHistory = new BrowserHistory("leetcode.com");
        browserHistory.visit("google.com");
        browserHistory.visit("facebook.com");
        browserHistory.visit("youtube.com");
        System.out.println(browserHistory.back(1));
    }
}

class BrowserHistory {

    ListNode curr;
    public BrowserHistory(String homepage) {
        curr = new ListNode(homepage);
    }
    
    public void visit(String url) {
        ListNode newNode = new ListNode(url);
        curr.next = newNode;
        newNode.prev = curr;
        curr = curr.next;
    }
    
    public String back(int steps) {
        while(steps > 0){
            if(curr.prev == null) break;
            curr = curr.prev;
            steps--;
        }

        return curr.url;
    }
    
    public String forward(int steps) {
        while(steps > 0){
            if(curr.next == null) break;
            curr = curr.next;
            steps--;
        }
        
        return curr.url;
    }
}

class ListNode{
    String url;
    ListNode next;
    ListNode prev;

    ListNode(String url){
        this.url = url;
    }
}

File: dsa/patterns/09_linkedlist/python/l1472_design_browser_history.py
"""
Problem: https://leetcode.com/problems/design-browser-history/

Idea:
Use a doubly linked list.
Maintain a current pointer.
- visit(): clear forward history and move current to new node
- back(): move left using prev pointer
- forward(): move right using next pointer
"""

class ListNode:
    def __init__(self, url: str):
        self.url = url
        self.prev = None
        self.next = None


class BrowserHistory:
    def __init__(self, homepage: str):
        self.curr = ListNode(homepage)

    def visit(self, url: str) -> None:
        new_node = ListNode(url)
        self.curr.next = new_node
        new_node.prev = self.curr
        self.curr = new_node

    def back(self, steps: int) -> str:
        while steps > 0 and self.curr.prev:
            self.curr = self.curr.prev
            steps -= 1
        return self.curr.url

    def forward(self, steps: int) -> str:
        while steps > 0 and self.curr.next:
            self.curr = self.curr.next
            steps -= 1
        return self.curr.url


# Example usage
if __name__ == "__main__":
    browserHistory = BrowserHistory("leetcode.com")
    browserHistory.visit("google.com")
    browserHistory.visit("facebook.com")
    browserHistory.visit("youtube.com")
    print(browserHistory.back(1))      # facebook.com
